### DISCUSSION TITLE
Adding custom context values

### DISCUSSION THREAD

**Discussion by EricDallAgnol** [atoti/atoti#561] [General]
Hi,

Currently, I have a PostProcessor, implemented under Java AP, that I would like to migrate to Atoti. But its behaviour is based on a **custom** context value, provided by the user in the UI.
Thus, my question is, how to reproduce that within Atoti ?
I would like to know (if possible) how to set a new custom context value, like having for instance something like that
`cube.shared_context["myCustomContextValue"] = 10`
And then, being able to get this value back on the evaluation of my measure.

Many thanks,

**Comment by bloussou** [comment]
Hello,
If I'm correctly understanding you are creating a new project using atoti and you would like to define a measure using context values?
If yes, this is not possible at the moment. What is your use case?

**Comment by fabiencelier** [comment]
An alternative to context values is to use a [parameter hierarchy from members](https://docs.atoti.io/latest/lib/atoti/atoti.cube.html#atoti.Cube.create_parameter_hierarchy_from_members) with a set of predefined values.

You can then use the value of this hierarchy in another measure and users will be able to pick one the values from the UI, but they won't be able to add custom values (not predefined)

**Comment by EricDallAgnol** [comment]
Hi,

Thank you both very much for your answers.
Correct @bloussou 

In fact, the use case is for Market Risk, we are receiving sensitivities data, related to two buckets fields (1D, 2D, ... 1Y... ) 
Our csv is like
sensi | timebucket_1 | timebucket_2

The aim is then to display the sensi values in "2D" to have an output like this : 
![image](https://user-images.githubusercontent.com/83015366/163429480-a82af56d-b60d-40b4-96dc-05b096ce0267.png)

However, for some measures using sensi, we want to have a specific granularity for the buckets, ie for one analysis we only want to display sensi for 1D, 2D, 1W and 1M while for another,  relevant buckets will be 1Y, 2Y, 5Y, and this for the same measure.

For that case using filters would work but, in addition, we have all some custom buckets like >= 1Y that are created "on runtime" based on the grid chosen, that involves repillaring work. 
Here the idea is to provide the grid name, through the context value, to the measure to : 
- first fetch the grid description (having the "standard" and custom buckets) from the name provided
- second to calculate custom buckets values on the fly (no need to anticipate all the possible combinations)


By any chance, is the ability to add new custom context values planned for future Atoti releases ?

Many thanks,

**Reply by rcolle** [reply]
Hi Eric,

As mentioned by Fabien, creating a parameter hierarchy from members (see [this](https://docs.atoti.io/latest/lib/atoti/atoti.cube.html#atoti.Cube.create_parameter_hierarchy_from_members)) is the new, more efficient way that replaces context values.
It creates a slicing hierarchy with all possible values, and lets the user choose the one they prefer (the grid name in your case) by simply selecting the member, instead of having to edit context values.

Nonetheless, this seems to be only half of your problem, since you then need to use that value to perform the required bucketing.
@alexis-bvq will follow up with you with a solution that fits the needs of your project.

Thanks!

**Reply by alexis-bvq** [reply]
Hello Eric,

I tried to replicat your use case in the simpliest way possible through this piece of code:

```python
import atoti as tt
import pandas as pd

session = tt.Session()

count=200
vector_size = 100

base_table = session.read_pandas(
    pd.DataFrame({
        "l1" : [f"a{i}" for i in range(count)],
        "sensi_vector": [[10.0 * i for j in range(vector_size)] for i in range(count)]
    }), 
    table_name="base_table",
    keys=["l1"]
)
          
cube = session.create_cube(base_table)   

# Create a hiearachy that will allow to decompose the initial vector by its initial buckets.
cube.create_parameter_hierarchy_from_members("initial_bucket", members=[f"d_{i}" for i in range(vector_size)], index_measure_name="vector index")
m, lvl, h = cube.measures, cube.levels, cube.hierarchies


# We are creating a bucket table that will contains the list of new buckets.
# For each value of the initial bucket will be link to the name of the targeted bucket as well as the bucket_scenario, which would be the name you were inserting in the context value.
list_bucket_names = ["bucket_scenario1", "bucket_scenario2"]
bucket_names = []
buckets_values = []
initial_buckets = []
for i in range(vector_size):
    bucket_names.append("initial_bucket")
    buckets_values.append(f"d_{i}")
    initial_buckets.append(f"d_{i}")
    
for j in range(len(list_bucket_names)):
    bucket_name = list_bucket_names[j]
    for i in range(vector_size):
        bucket_names.append(bucket_name)
        buckets_values.append(f"m_{i % (j + 4)}")
        initial_buckets.append(f"d_{i}")

bucket_table = session.read_pandas(
    pd.DataFrame({
        "initial_bucket": initial_buckets,
        "bucket_scenario": bucket_names,
        "new_bucket": buckets_values 
    }), 
    table_name="new_buckets", 
    keys=["initial_bucket", "bucket_scenario", "new_bucket"]
)

# We join this table to the initial bucket table created with the parameter hierarchy function.
session.tables["initial_bucket"].join(bucket_table, mapping={"initial_bucket": "initial_bucket"})


# We created the measure with the assumptions that the vector was already interpolated in the ETL, but you could obviously use the udaf function passed in the ticket.
m["Vector Expand"] = tt.agg.sum(m["sensi_vector.SUM"][m["vector index"]], scope=tt.scope.origin(lvl["initial_bucket"]))
h["initial_bucket"].slicing = False
h["initial_bucket"].visible = False
h["new_bucket"].slicing = False
h["bucket_scenario"].slicing = True

# Will return the result of the query at the initial bucket description.
cube.query(m["Vector Expand"], levels=[lvl["new_bucket"]], condition=lvl["bucket_scenario"] == "initial_bucket")

# Will return the result with the new bucket of scenario one.
cube.query(m["Vector Expand"], levels=[lvl["new_bucket"]], condition=lvl["bucket_scenario"] == list_bucket_names[1])
```

So here you don't have the context value anymore but it is replaced by the `bucket scenario` slicing hierarchy, and instead of having hardcoded `map` within the java code for each different context value, you just have to feed the `bucket_table` with the different mapping you need.

I hope it is clear enough, if not we can obviously make another session together so that I can explain it in more details.

Here I have done it with only one expansion, you could do it a second time replicating the same flow, and then create a `where` measure to take one expansion or the other accordingly to which timeSeries expansion you are looking at.

**Comment by EricDallAgnol** [comment]
Hello,

Many thanks Alexis, I will have a look at that and let you know !

In any case it is very appreciable to have such a reactivity from the Atoti team!!! Great job!
